kube/src/lib.rs

```diff
@@ -431,6 +434,103 @@ mod test {
         Ok(())
     }
 
+    #[tokio::test]
+    #[ignore] // needs cluster (fetches api resources, and lists all)
+    #[cfg(all(feature = "derive"))]
+    async fn derived_resources_by_stability_discoverable() -> Result<(), Box<dyn std::error::Error>> {
+        use crate::{
+            discovery::{self, Discovery},
+            runtime::wait::{await_condition, conditions},
+        };
+        use kube_core::crd::merge_crds;
+
+        mod v1alpha1 {
+            use super::*;
+
+            #[derive(CustomResource, Deserialize, Serialize, Clone, Debug, JsonSchema)]
+            #[kube(
+                group = "kube.rs",
+                version = "v1alpha1",
+                kind = "TestLowVersionCr",
+                namespaced
+            )]
+            #[kube(crates(kube_core = "crate::core"))] // for dev-dep test structure
+            pub struct TestLowVersionCrSpec {}
+        }
+
+        mod v1 {
+            use super::*;
+
+            #[derive(CustomResource, Deserialize, Serialize, Clone, Debug, JsonSchema)]
+            #[kube(group = "kube.rs", version = "v1", kind = "TestCr", namespaced)]
+            #[kube(crates(kube_core = "crate::core"))] // for dev-dep test structure
+            pub struct TestCrSpec {}
+        }
+
+        mod v2alpha1 {
+            use super::*;
+
+            #[derive(CustomResource, Deserialize, Serialize, Clone, Debug, JsonSchema)]
+            #[kube(group = "kube.rs", version = "v2alpha1", kind = "TestCr", namespaced)]
+            #[kube(crates(kube_core = "crate::core"))] // for dev-dep test structure
+            pub struct TestCrSpec {}
+        }
+
+        async fn apply_crd(
+            client: Client,
+            name: &str,
+            crd: CustomResourceDefinition,
+        ) -> Result<(), Box<dyn std::error::Error>> {
+            let crds: Api<CustomResourceDefinition> = Api::all(client.clone());
+            let ssapply = PatchParams::apply("kube").force();
+            crds.patch(name, &ssapply, &Patch::Apply(&crd)).await?;
+            let establish = await_condition(crds.clone(), name, conditions::is_crd_established());
+            let _ = tokio::time::timeout(std::time::Duration::from_secs(10), establish).await?;
+            Ok(())
+        }
+
+        let client = Client::try_default().await?;
+        let test_lowversion_crd = v1alpha1::TestLowVersionCr::crd();
+        let testcrd_v1 = v1::TestCr::crd();
+        let testcrd_v2alpha1 = v2alpha1::TestCr::crd();
+        let all_crds = vec![testcrd_v1.clone(), testcrd_v2alpha1.clone()];
+
+        apply_crd(client.clone(), "testlowversioncrs.kube.rs", test_lowversion_crd).await?;
+        apply_crd(client.clone(), "testcrs.kube.rs", merge_crds(all_crds, "v1")?).await?;
+
```

Comment:
I think we actually need an extra 5s sleep here after wating for established conditions. Just because a resource has been established on the apiserver does not mean it will show up in discovery (and there's no way to wait for it it except for retrying the discovery).

Reply:
sleep had been added in app_crd function.

Reply:
Can we put the sleep after the `apply_crd` calls? Otherwise this test takes 20s :-)

Reply:
ok

Reply:
It is kind difficult to move the sleep out of `apply_crd` func. Any help?

Reply:
The sleep that is there is awaiting the condition (and needs to be there).

You would need something like `tokio::time::sleep(std::time::Duration::from_secs(5)).await; outside.

Reply:
to give a unit test in apigroup.rs, so we don't need this test now.